code/app/src/main/java/com/example/bread/view/AddMoodEventActivity.java

```diff
@@ -0,0 +1,181 @@
+        if (trigger.isEmpty()) trigger = null;
+
+        // Get selected emotional state (required)
+        MoodEvent.EmotionalState emotionalState = (MoodEvent.EmotionalState) emotionalStateSpinner.getSelectedItem();
+        if (emotionalState == null) {
+            Toast.makeText(this, "Please select a mood!", Toast.LENGTH_SHORT).show();
+            return;
+        }
+
+        // Get social situation (optional)
+        MoodEvent.SocialSituation socialSituation = (MoodEvent.SocialSituation) socialSituationSpinner.getSelectedItem();
+
+        // Handle location (optional, based on checkbox)
+        AtomicReference<Map<String, Object>> geoInfo = new AtomicReference<>();
+        if (locationCheckbox.isChecked()) {
+            if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
+                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, LOCATION_PERMISSION_REQUEST_CODE);
+                Toast.makeText(this, "Location permission required. Please grant permission and try again.", Toast.LENGTH_SHORT).show();
+                return;
+            }
+
+            String finalEventTitle = eventTitle;
+            String finalReason = reason;
+            String finalTrigger = trigger;
+            fusedLocationClient.getLastLocation().addOnSuccessListener(this, location -> {
+                if (location != null) {
+                    geoInfo.set(new MoodEvent().generateGeoInfo(location)); // Reuse MoodEvent's method
+                    saveMoodEventWithLocation(geoInfo.get(), finalEventTitle, finalReason, emotionalState, socialSituation, finalTrigger);
+                } else {
+                    Toast.makeText(this, "Unable to get location. Please enable GPS and try again.", Toast.LENGTH_SHORT).show();
+                }
+            });
+        } else {
+            saveMoodEventWithLocation(null, eventTitle, reason, emotionalState, socialSituation, trigger);
+        }
+    }
+
+    private void saveMoodEventWithLocation(Map<String, Object> geoInfo, String eventTitle, String reason,
+                                           MoodEvent.EmotionalState emotionalState, MoodEvent.SocialSituation socialSituation,
+                                           String trigger) {
+        // Fetch the current participant from Firebase using ParticipantRepository
+        participantRepository.fetchParticipant(getCurrentUsername(), participant -> {
+            if (participant == null) {
+                Toast.makeText(this, "User not found. Please log in or register!", Toast.LENGTH_SHORT).show();
+                return;
+            }
+
+            // Get the participant's username or use a unique identifier (e.g., document ID)
+            String username = participant.getUsername();
+            DocumentReference participantRef = FirebaseFirestore.getInstance().collection("participants").document(username);
```

Comment:
Again, refrain from using firebase direct. Use the already available static `FirebaseService` written if you need to use it. Otherwise always use ParticipantRepository as mush as possible If you need a method that doesn't exist in it "Add it".